### ISSUE TITLE
Db connection is hung when there is a network change

### ISSUE THREAD

**Issue by tuliren** [LiveRamp/jack#227]
A db connection will hang when there is a network change. This is because we don't enforce timeout at any level. For detailed explanation, please refer to [this article](https://www.cubrid.org/blog/understanding-jdbc-internals-and-timeout-configuration).

There are several ways to fix it:
1. Specify socket timeout. This is easy to implement (modify [connection string](https://github.com/LiveRamp/jack/blob/master/jack-core/src/com/rapleaf/jack/DatabaseConnection.java#L92)), but it is hard to find the optimal value, or to use special setting for legitimate long-running query.
2. Enforce timeout on transactor. This requires a refactoring of the transactor implementation (relevant PR: https://github.com/LiveRamp/jack/pull/210).
3. Call [`isValid`](https://docs.oracle.com/javase/7/docs/api/java/sql/Connection.html#isValid(int)) before checking out a connection from the pool (modify [this method](https://github.com/LiveRamp/jack/blob/master/jack-core/src/com/rapleaf/jack/transaction/DbPoolManager.java#L167)). This approach will only partially solve this problem. Queries that are already running while there is a network change will still suffer from this issue. In addition, currently `IDb` methods are implemented in each implementation, while many of the methods are the same and can be extracted to a base class. Should we choose this approach, we should consider performing this refactoring as well.

**Comment by joshk0** [comment]
Can tcp keepalives be used?

This would prevent closing a long running transaction, as long as the
server is responding at a low level, it does not need to be returning query
results.

**Comment by sidoh** [comment]
@joshk0 - I had the same idea, but I think you have the firmer understanding here.  It seems like this kind of thing is exactly what SO_KEEPALIVE is intended to solve for.  It's also just a JDBC parameter:

https://dev.mysql.com/doc/connector-j/5.1/en/connector-j-reference-configuration-properties.html

**Comment by joshk0** [comment]
Yes, I'd suggest trying that first. You can even collaborate with us to
simulate a network outage on a specific VM whenever you want.
